### ISSUE TITLE
--restart args not implement 

### ISSUE THREAD

**Issue by firmanjml** [cockpit-project/cockpit-podman#817]
It would be nice to have --restart options on the gui so user can choose to set auto restart, restart on crash or etc.

https://stackoverflow.com/a/63795225

**Comment by garrett** [comment]
Totally agreed!

When we were working on the redesign, Podman did not have this feature (in a way that worked). We would've had to add the systemd unit integration and track that ourselves. However, Podman did add this feature a few months ago, and we definitely want to support it!

I added this to our design document today, tacked on to the notes I copy/pasted there today @ https://github.com/cockpit-project/cockpit/discussions/16059#discussioncomment-1663235

To implement it properly, we'll need to make sure the system service is running. I also don't know how that will work for a rootless container — will it run the containers after boot (wanted) or only after a user logs in (mostly pointless)? We'll need to make sure it works for both system and user containers, or we'll need to figure out a workaround otherwise.

**Comment by garrett** [comment]
To add to that: We'd need to make sure `podman-restart.service` is enabled, so it'll run on boot to re-start applicable containers. (As far as I know, it starts on boot, runs the containers it should, and then exits. It's not a daemon that always runs.)

If it's not enabled by default, we'd need to enable it ourselves and should probably inform the user that it will be enabled.

Again, I don't know what happens with rootless user containers, but we should investigate further.

**Comment by garrett** [comment]
Note: Using `podman-restart.service` is different from the systemd unit generation (which we'd have to manage ourselves otherwise). It arrived in https://podman.io/releases/2021/08/31/podman-release-v3.3.0.html

**Comment by jelly** [comment]
> Note: Using `podman-restart.service` is different from the systemd unit generation (which we'd have to manage ourselves otherwise). It arrived in https://podman.io/releases/2021/08/31/podman-release-v3.3.0.html

Tested it locally and it only works with system containers sadly, however it should in theory work for a user session:

This is what the systemd unit executes.
```
[admin@fedora-34-127-0-0-2-2201 ~]$ podman start --all --filter restart-policy=always
810904806fae16a08e41397c61bb0b9e3e5814de1f5467f762a546174e73efe7
```

Container starts.
```
[admin@fedora-34-127-0-0-2-2201 ~]$ podman ps
CONTAINER ID  IMAGE                             COMMAND     CREATED         STATUS            PORTS       NAMES
810904806fae  docker.io/grafana/grafana:latest              23 minutes ago  Up 3 seconds ago              rebootproof
```

So podman-restart would need to support [an user identifier](https://man.archlinux.org/man/systemd.unit.5#SPECIFIERS) so we can ask a user to enable `systemd enable podman-restart@jelle`.

**Comment by jelly** [comment]
I've made a quick implementation, but this needs some design:

When on-failure is set, the number of failures can be specified. However I am not sure how we can put them on the same line while having a label to describe the TextInput.
![image](https://user-images.githubusercontent.com/67428/142657583-bc5da2fb-e674-4fd1-b707-d388eec75163.png)

When setting always, the user might want the container to be restarted on boot, this requires enabling a systemd unit, how do we integrate it in the dialog if it's not already enabled?
![image](https://user-images.githubusercontent.com/67428/142657835-207458ba-9e7d-49e4-9869-b086c031e46f.png)

**Comment by martinpitt** [comment]
> this requires enabling a systemd unit, how do we integrate it in the dialog if it's not already enabled?

IMHO, don't. It's an internal implementation detail, and honestly podman should enable it by itself on demand. I suggest that you just enable the system unit if the restart policy is set accordingly (but *never* automatically disable the unit!).

The user unit is a different beast -- if it makes sense to start "always" containers the first time the user logs in, then merely enabling the user unit is fine. But IMHO that's unintuitive. We certainly should *not* do the `loginctl enable-linger` automatically, as this is very intrusive, very unexpected, often does not even work, could be a giant security problem, and it's hard to explain the consequences.

IMHO we should just not offer the restart policy for user containers -- if you need that level of control, use a system container.

**Comment by jelly** [comment]
> > this requires enabling a systemd unit, how do we integrate it in the dialog if it's not already enabled?
> 
> IMHO, don't. It's an internal implementation detail, and honestly podman should enable it by itself on demand. I suggest that you just enable the system unit if the restart policy is set accordingly (but _never_ automatically disable the unit!).

Sounds good!

> The user unit is a different beast -- if it makes sense to start "always" containers the first time the user logs in, then merely enabling the user unit is fine. But IMHO that's unintuitive. We certainly should _not_ do the `loginctl enable-linger` automatically, as this is very intrusive, very unexpected, often does not even work, could be a giant security problem, and it's hard to explain the consequences.
> 
> IMHO we should just not offer the restart policy for user containers -- if you need that level of control, use a system container.

I agree it's more of an issue with the user session, not sure how often that is a use case. But it would be worthwhile to maybe link to the documentation? Although it seems it's a bit lacking from the podman side. The [man page](https://docs.podman.io/en/latest/markdown/podman-run.1.html?highlight=podman-run#restart-policy) does not include a section about `podman-restart.service` or having to enable linger for user containers

**Comment by garrett** [comment]
I messed with some JSX code on PatternFly.com, in the grid form area to create this mockup:

![Screenshot 2021-11-30 at 14-45-48 PatternFly 4 • Form](https://user-images.githubusercontent.com/10246/144058679-78849fd9-b65a-4b96-aeec-ffcc672b9450.png)

It's using [grid form](https://www.patternfly.org/v4/components/form#grid-form) with a [form select](https://www.patternfly.org/v4/components/form-select) for the restart policy and a maximum retries as a [number input](https://www.patternfly.org/v4/components/number-input). It's disabled in this screenshot, as it isn't properly hooked up, but would be enabled in the real thing. The default would be the default number of tries that Podman gives.

 If the restart policy is something that doesn't need retries, just don't show the maximum retries (but keep the same width):

![Screenshot 2021-11-30 at 14-47-59 PatternFly 4 • Form](https://user-images.githubusercontent.com/10246/144059415-1cab9f7b-2220-49ae-a3ae-87fcb54a0793.png)



The JSX I cobbled together is a mess and was just for an example. It was just enough to get it to not have an error on the website and show something:

```jsx
import React from 'react';
import { Form, FormGroup, FormSelect, FormSelectOption, TextInput, Button, Grid, GridItem, NumberInput } from '@patternfly/react-core';
import HelpIcon from '@patternfly/react-icons/dist/esm/icons/help-icon';

class SimpleForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value1: '',
      value2: '',
      value3: ''
    };
    this.handleTextInputChange1 = value1 => {
      this.setState({ value1 });
    };
    this.handleTextInputChange2 = value2 => {
      this.setState({ value2 });
    };
    this.handleTextInputChange3 = value3 => {
      this.setState({ value3 });
    };
  }

  render() {
    const { value1, value2, value3 } = this.state;

    return (
      <Form isHorizontal>
        <Grid hasGutter md={6}>
          <GridItem span={12}>
            <FormGroup label="Full name" isRequired fieldId="grid-form-name-01" helperText="Include your middle name if you have one.">
              <TextInput
                isRequired
                type="text"
                id="grid-form-name-01"
                name="grid-form-name-01"
                aria-describedby="grid-form-name-01-helper"
                value={value1}
                onChange={this.handleTextInputChange1}
              />

            </FormGroup>
          </GridItem>
          <FormGroup label="Restart Policy"  fieldId="grid-form-email-01">
          
                                <FormSelect value="On failure" aria-label="FormSelect Input">
          <FormSelectOption key={1} value="On failure" label="On failure" />
      </FormSelect>

          </FormGroup>
          <FormGroup label="Maximum retries" fieldId="grid-form-number-01">
                  <NumberInput
        value={5}
        inputName="input"
        inputAriaLabel="number input"
        minusBtnAriaLabel="minus"
        plusBtnAriaLabel="plus"
      />

          </FormGroup>
        </Grid>
      </Form>
    );
  }
}
```

**Comment by garrett** [comment]
>> this requires enabling a systemd unit, how do we integrate it in the dialog if it's not already enabled?
>
> IMHO, don't. It's an internal implementation detail, and honestly podman should enable it by itself on demand.

This is what I suggested from the beginning. I'm happy to see @martinpitt also suggesting this.

As for users, we could either not show the restart option (which could be a bit confusing) or show the controls and when it's set, show a warning that it requires linger to be enabled. Ordinarily, I'd just say that we should make it easy for someone to set the linger, except for @martinpitt saying that it is _"very intrusive, very unexpected, often does not even work, could be a giant security problem, and it's hard to explain the consequences"._.. so _if_ we do show the option for users, we should show a warning and might want to elaborate in a (?) popover on the warning.

**Comment by jelly** [comment]
https://github.com/cockpit-project/cockpit-podman/issues/364 is basically the same as this issue.
